client/cypress/end-to-end/offline-batch-data-entry-tests.spec.js

```diff
@@ -0,0 +1,296 @@
+
+    cy.wait(100) // gets stuck in an infinite loop without a 100ms wait here
+    cy.findByText('Next').click()
+    cy.get('input[name="contests[0].name"]').type('Contest')
+    cy.get('input[name="contests[0].choices[0].name"]').type('A')
+    cy.get('input[name="contests[0].choices[0].numVotes"]').type('300')
+    cy.get('input[name="contests[0].choices[1].name"]').type('B')
+    cy.get('input[name="contests[0].choices[1].numVotes"]').type('100')
+    cy.get('input[name="contests[0].totalBallotsCast"]').type('400')
+    cy.findByText('Select Jurisdictions').click()
+    cy.findByLabelText('Death Star').check({ force: true })
+    cy.findByText('Save & Next').click()
+    cy.findAllByText('Opportunistic Contests').should('have.length', 2)
+    cy.findByText('Save & Next').click()
+    cy.get('#state').select('AL')
+    cy.get('input[name=electionName]').type(`Test Election`)
+    cy.get('#risk-limit').select('10')
+    cy.get('input[name=randomSeed]').type('543210')
+    cy.findByText('Save & Next').click()
+    cy.findAllByText('Review & Launch').should('have.length', 2)
+    cy.logout(auditAdmin)
+    cy.loginJurisdictionAdmin(jurisdictionAdmin)
+    cy.findByText(`Jurisdictions - TestAudit${id}`)
+      .siblings('button')
+      .click()
+    cy.fixture('CSVs/manifest/ballot_polling_manifest.csv').then(
+      fileContent => {
+        cy.get('input[type="file"]')
+          .first()
+          .attachFile({
+            fileContent: fileContent.toString(),
+            fileName: 'ballot_polling_manifest.csv',
+            mimeType: 'csv',
+          })
+      }
+    )
+    cy.findByText('Upload File').click()
+    cy.contains('Upload successfully completed')
+    cy.logout(jurisdictionAdmin)
+    cy.loginAuditAdmin(auditAdmin)
+    cy.findByText(`TestAudit${id}`).click()
+    cy.findByText('Review & Launch').click()
+    cy.findAllByText('Review & Launch').should('have.length', 2)
+
+    // add custom sample size to be same as total ballots cast
+    cy.findByText('Enter your own sample size (not recommended)').click()
+    cy.findByRole('spinbutton').type('400').blur()
+
+    cy.findByRole('button', { name: 'Launch Audit' })
+      .should('be.enabled')
+      .click()
+    cy.findAllByText('Launch Audit').spread((firstButton, secondButton) => {
+      secondButton.click()
+    })
+    cy.findByRole('heading', { name: 'Audit Progress' })
+    cy.logout(auditAdmin)
+    cy.loginJurisdictionAdmin(jurisdictionAdmin)
+    cy.findByText(`Jurisdictions - TestAudit${id}`)
+      .siblings('button')
+      .click()
+    cy.contains('Number of Audit Boards')
+    cy.findByText('Save & Next').click()
+    cy.contains('No batches added. Add your first batch below.')
+
+    // check validation message when no fields are filled up
+    cy.findByRole('button', { name: /Add batch/ }).click()
+    cy.findByRole('button', {name: 'Save Batch'}).click()
+    cy.contains('Please fill in the empty fields above before saving this batch.')
```

Comment:
i'm really glad you added an end to end test for this flow - that's definitely something we were missing. at the same time, the cypress tests are already the longest running part of CI. so ideally we would only test the bare minimum in the cypress test and pull the rest out into unit tests. for example, this validation message is entirely frontend, so it could be tested with a unit test